### ISSUE TITLE
SQLAlchemy @hybrid_property decorator not supported

### ISSUE THREAD

**Issue by tdamsma** [PyCQA/pylint#1061]
I use SQLAlchemy, which has the `@hybrid_property` decorator, as an alternative to to the`@property` decorator. 

If one uses the `@property`, this is detected by the `decorated_with_property` function in `pylint.checkers.utils`, but this function does not detect the use of `@hybrid_property`, leading to false positives for [E0202](https://github.com/pchomik/linter-python-doc/blob/master/errors/pylint.md#e0202)

Not sure if this falls in the scope of what pylint wants to achieve, but I would appreciate support for this SQLAlchemy specific construct.

**Comment by PCManticore** [comment]
Hi, sorry for taking so long to reply.

Can you provide an example where this causes you problems? I think we already have a mechanism for overriding the builtin decorators that pylint can understand (property-classes if I am remembering correctly).

**Comment by tdamsma** [comment]
See below for a typical sqlalchemy use case that throws an `E0202` error. If you replace `@hybrid_property` with '@property` there is no error.

```python
from sqlalchemy import Column, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property

Base = declarative_base()


class Interval(Base):
    __tablename__ = 'interval'

    id = Column(Integer, primary_key=True)
    start = Column(Integer, nullable=False)
    end = Column(Integer, nullable=False)

    def __init__(self, start, end):
        self.start = start
        self.length = 2

    @hybrid_property
    def length(self):
        return self.end - self.start

    @length.setter
    def length(self, value):
        self.end = self.start + value
```

**Comment by m3talstorm** [comment]
I have basically the same issue, but I get a E1101 (no-member) because I'm returning an object, which I then access a member on.

**Comment by AWhetter** [comment]
`hybrid_property` doesn't inherit from `property`, so we would have to look for a [property-like interface](https://docs.python.org/3.6/howto/descriptor.html#properties) to support duck typing.

Although my guess is that `_is_property_decorator` (https://github.com/PyCQA/pylint/blob/fbdf5ba095ebbbf04db12f6c3e88e21152a30507/pylint/checkers/utils.py#L500) cannot infer `hybrid_property`, and so it assumes that it isn't a property.
Not really sure how we can go about this without adding a configurable whitelist of property names.

**Comment by domruf** [comment]
+1 for property whitelist

**Comment by craiga** [comment]
Flask has introduced [`flask.helpers.locked_cached_property`](https://github.com/pallets/flask/blob/master/flask/helpers.py#L840) which similarly decorates methods so they behave as properties:

```python
    @locked_cached_property
    def logger(self):
        return create_logger(self)
```

As I read the documentation, you _should_ be able to set `property-classes` to tell pylint about these decorators, but I can't get it to work.

```
# List of decorators that produce properties, such as abc.abstractproperty. Add
# to this list to register other decorators that produce valid properties.
property-classes = abc.abstractproperty, flask.helpers.locked_cached_property
```

```
$ pipenv run pylint unixtimestamp.py 
Loading .env environment variables…
Using config file /Users/craiga/unixtimesta.mp/.pylintrc
************* Module unixtimestamp
E: 19, 0: Method 'logger' has no 'addHandler' member (no-member)
E: 20, 0: Method 'logger' has no 'setLevel' member (no-member)
E: 54, 8: Method 'logger' has no 'info' member (no-member)
E: 87, 8: Method 'logger' has no 'info' member (no-member)
E:183, 8: Method 'logger' has no 'info' member (no-member)
E:223, 4: Method 'logger' has no 'info' member (no-member)

------------------------------------------------------------------
Your code has been rated at 7.27/10 (previous run: 7.27/10, +0.00)
```

**Comment by AWhetter** [comment]
Providing the module that contains the property decorator is discoverable and inferable by pylint, this issue should now be resolved.
For all other cases, unfortunately configuration options are not shareable between checkers. The checker that defines `property-classes` is different to the one that raises E0202, so E0202 cannot use the whitelist of property names. This issue is being resolved as part of the per directory config project though, so I'm closing this issue with the partial fix in place.

**Comment by chaoflow** [comment]
@AWhetter The case of sqlalchemy's hybrid_property (and hybrid_method) is more complex than what has been discussed here: It behaves differently when used on class and instance level see https://docs.sqlalchemy.org/en/13/orm/extensions/hybrid.html#module-sqlalchemy.ext.hybrid

Do you see how these could be supported in pylint?

**Comment by chaoflow** [comment]
See here an example how different instance-level and class-level behaviour is defined: https://docs.sqlalchemy.org/en/13/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior

**Comment by kevlarr** [comment]
Having same issue here with `@hybrid_property` resulting in `E1101` despite setting `property-classes=abc.abstractproperty, sqlalchemy.ext.hybrid.hybrid_property`

**Comment by lesinigo** [comment]
And we're having the same issue with `W0143 comparison-with-callable` despite having configured in out `pyproject.toml`:

```toml
[tool.pylint.basic]
property-classes = ["abc.abstractproperty", "sqlalchemy.ext.hybrid.hybrid_property"]
```
